pkg/namespacecheck/ownership_test.go

```diff
@@ -0,0 +1,107 @@
+package namespacecheck
+
+import (
+	"encoding/json"
+	"fmt"
+	v1 "github.com/openshift-eng/ci-test-mapping/pkg/api/types/v1"
+	"github.com/openshift-eng/ci-test-mapping/pkg/config"
+	"github.com/openshift-eng/ci-test-mapping/pkg/registry"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"os"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func TestNoOverlap(t *testing.T) {
+	defaultRegistry := registry.NewComponentRegistry()
+
+	componentNameToNamespaces := map[string][]string{}
+	namespacesToComponentNames := map[string][]string{}
+
+	for _, component := range defaultRegistry.Components {
+		// All types currently conform.  If any don't in the future, find a way to handle them.
+		// I recommend not skipping them.  Probably force conformance
+		componentValuePointer := reflect.ValueOf(component)
+		componentValue := componentValuePointer.Elem()
+		defaultComponentValuePointer := componentValue.FieldByName("Component")
+		asInterface := defaultComponentValuePointer.Interface()
+
+		defaultComponent := asInterface.(*config.Component)
+		for _, matcher := range defaultComponent.Matchers {
+			for _, namespace := range matcher.Namespaces {
+				namespacesToComponentNames[namespace] = append(namespacesToComponentNames[namespace], defaultComponent.Name)
+				componentNameToNamespaces[defaultComponent.Name] = append(componentNameToNamespaces[defaultComponent.Name], namespace)
+			}
+		}
+
+	}
+
+	for namespace, components := range namespacesToComponentNames {
+		if len(components) > 1 {
+			t.Fatalf("namespace/%v is claimed by more than one component: %v", namespace, strings.Join(components, ", "))
```

Comment:
This is handled in `make mapping`, if you end up with a test claimed by two components, it fails and tells you why (and it does this in the PR presubmit too, blocking merge).  Is all this namespacechecking needed?

Reply:
> This is handled in `make mapping`, if you end up with a test claimed by two components, it fails and tells you why (and it does this in the PR presubmit too, blocking merge). Is all this namespace checking needed?

The `make mapping` only covers whether tests have well known priority.  You can still have overlapping namespaces with different priorities.

Reply:
Well it ensures that two or more components don't claim ownership of the same test. 

This test says that a namespace can only belong to one component.  That's different than how the other matchers work.

I'd expect to be able to do this.

In the monitoring component:
```
{
           IncludeAny: []string{"apigroup:monitoring.coreos.com"},
           Namespaces: []string{"openshift"},
           Priority: 1,
}
```

In some other component:
```
{
           Namespaces: []string{"openshift"},
           Priority: 1,
}
```

Which would split ownership of the openshift namespace.

If that's not desirable, maybe the Namespaces belongs outside of the matchers, and as part of the general component config.

Reply:
> If that's not desirable, maybe the Namespaces belongs outside of the matchers, and as part of the general component config (like Operators)

It is not desireable.  Are there currently any matching capabilities outside the matcher?  This test prevents the situation.

Reply:
Yes, Operators